<!DOCTYPE html>
<html>
<head><title>R: Immunoglobulin Clonal Lineage and Diversity Analysis</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> Immunoglobulin Clonal Lineage and Diversity Analysis
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;alakazam&rsquo; version 1.2.1</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
</ul>

<h2>Help Pages</h2>


<p style="text-align: center;">
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#E">E</a>
<a href="#G">G</a>
<a href="#I">I</a>
<a href="#J">J</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#P">P</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#W">W</a>
</p>


<h2><a name="A">-- A --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="ABBREV_AA.html">ABBREV_AA</a></td>
<td>Amino acid abbreviation translations</td></tr>
<tr><td style="width: 25%;"><a href="AbundanceCurve-class.html">AbundanceCurve</a></td>
<td>S4 class defining a clonal abundance curve</td></tr>
<tr><td style="width: 25%;"><a href="AbundanceCurve-class.html">AbundanceCurve-class</a></td>
<td>S4 class defining a clonal abundance curve</td></tr>
<tr><td style="width: 25%;"><a href="AbundanceCurve-class.html">AbundanceCurve-method</a></td>
<td>S4 class defining a clonal abundance curve</td></tr>
<tr><td style="width: 25%;"><a href="alakazam.html">alakazam</a></td>
<td>The alakazam package</td></tr>
<tr><td style="width: 25%;"><a href="aliphatic.html">aliphatic</a></td>
<td>Calculates the aliphatic index of amino acid sequences</td></tr>
<tr><td style="width: 25%;"><a href="alphaDiversity.html">alphaDiversity</a></td>
<td>Calculate clonal alpha diversity</td></tr>
<tr><td style="width: 25%;"><a href="aminoAcidProperties.html">aminoAcidProperties</a></td>
<td>Calculates amino acid chemical properties for sequence data</td></tr>
</table>

<h2><a name="B">-- B --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="baseTheme.html">baseTheme</a></td>
<td>Standard ggplot settings</td></tr>
<tr><td style="width: 25%;"><a href="buildPhylipLineage.html">buildPhylipLineage</a></td>
<td>Infer an Ig lineage using PHYLIP</td></tr>
<tr><td style="width: 25%;"><a href="bulk.html">bulk</a></td>
<td>Calculates the average bulkiness of amino acid sequences</td></tr>
</table>

<h2><a name="C">-- C --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="calcCoverage.html">calcCoverage</a></td>
<td>Calculate sample coverage</td></tr>
<tr><td style="width: 25%;"><a href="calcDiversity.html">calcDiversity</a></td>
<td>Calculate the diversity index</td></tr>
<tr><td style="width: 25%;"><a href="ChangeoClone-class.html">ChangeoClone</a></td>
<td>S4 class defining a clone</td></tr>
<tr><td style="width: 25%;"><a href="ChangeoClone-class.html">ChangeoClone-class</a></td>
<td>S4 class defining a clone</td></tr>
<tr><td style="width: 25%;"><a href="charge.html">charge</a></td>
<td>Calculates the net charge of amino acid sequences.</td></tr>
<tr><td style="width: 25%;"><a href="checkColumns.html">checkColumns</a></td>
<td>Check data.frame for valid columns and issue message if invalid</td></tr>
<tr><td style="width: 25%;"><a href="collapseDuplicates.html">collapseDuplicates</a></td>
<td>Remove duplicate DNA sequences and combine annotations</td></tr>
<tr><td style="width: 25%;"><a href="combineIgphyml.html">combineIgphyml</a></td>
<td>Combine IgPhyML object parameters into a dataframe</td></tr>
<tr><td style="width: 25%;"><a href="countClones.html">countClones</a></td>
<td>Tabulates clones sizes</td></tr>
<tr><td style="width: 25%;"><a href="countGenes.html">countGenes</a></td>
<td>Tabulates V(D)J allele, gene or family usage.</td></tr>
<tr><td style="width: 25%;"><a href="countPatterns.html">countPatterns</a></td>
<td>Count sequence patterns</td></tr>
<tr><td style="width: 25%;"><a href="cpuCount.html">cpuCount</a></td>
<td>Available CPU cores</td></tr>
</table>

<h2><a name="D">-- D --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="DEFAULT_COLORS.html">DEFAULT_COLORS</a></td>
<td>Default colors</td></tr>
<tr><td style="width: 25%;"><a href="DiversityCurve-class.html">DiversityCurve</a></td>
<td>S4 class defining a diversity curve</td></tr>
<tr><td style="width: 25%;"><a href="DiversityCurve-class.html">DiversityCurve-class</a></td>
<td>S4 class defining a diversity curve</td></tr>
<tr><td style="width: 25%;"><a href="DiversityCurve-class.html">DiversityCurve-method</a></td>
<td>S4 class defining a diversity curve</td></tr>
<tr><td style="width: 25%;"><a href="DEFAULT_COLORS.html">DNA_COLORS</a></td>
<td>Default colors</td></tr>
<tr><td style="width: 25%;"><a href="IUPAC_CODES.html">DNA_IUPAC</a></td>
<td>IUPAC ambiguous characters</td></tr>
</table>

<h2><a name="E">-- E --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="EdgeTest-class.html">EdgeTest</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="EdgeTest-class.html">EdgeTest-class</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="EdgeTest-class.html">EdgeTest-method</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="estimateAbundance.html">estimateAbundance</a></td>
<td>Estimates the complete clonal relative abundance distribution</td></tr>
<tr><td style="width: 25%;"><a href="Example10x.html">Example10x</a></td>
<td>Small example 10x Genomics Ig V(D)J sequences from CD19+ B cells isolated from PBMCs of a healthy human donor. Down-sampled from data provided by 10x Genomics under a Creative Commons Attribute license, and processed with their Cell Ranger pipeline.</td></tr>
<tr><td style="width: 25%;"><a href="ExampleDb.html">ExampleDb</a></td>
<td>Example AIRR database</td></tr>
<tr><td style="width: 25%;"><a href="ExampleDbChangeo.html">ExampleDbChangeo</a></td>
<td>Example Change-O database</td></tr>
<tr><td style="width: 25%;"><a href="ExampleTrees.html">ExampleTrees</a></td>
<td>Example Ig lineage trees</td></tr>
<tr><td style="width: 25%;"><a href="extractVRegion.html">extractVRegion</a></td>
<td>Extracts FWRs and CDRs from IMGT-gapped sequences</td></tr>
</table>

<h2><a name="G">-- G --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="getAAMatrix.html">getAAMatrix</a></td>
<td>Build an AA distance matrix</td></tr>
<tr><td style="width: 25%;"><a href="getSegment.html">getAllele</a></td>
<td>Get Ig segment allele, gene and family names</td></tr>
<tr><td style="width: 25%;"><a href="getSegment.html">getChain</a></td>
<td>Get Ig segment allele, gene and family names</td></tr>
<tr><td style="width: 25%;"><a href="getDNAMatrix.html">getDNAMatrix</a></td>
<td>Build a DNA distance matrix</td></tr>
<tr><td style="width: 25%;"><a href="getSegment.html">getFamily</a></td>
<td>Get Ig segment allele, gene and family names</td></tr>
<tr><td style="width: 25%;"><a href="getSegment.html">getGene</a></td>
<td>Get Ig segment allele, gene and family names</td></tr>
<tr><td style="width: 25%;"><a href="getSegment.html">getLocus</a></td>
<td>Get Ig segment allele, gene and family names</td></tr>
<tr><td style="width: 25%;"><a href="getMRCA.html">getMRCA</a></td>
<td>Retrieve the first non-root node of a lineage tree</td></tr>
<tr><td style="width: 25%;"><a href="getPathLengths.html">getPathLengths</a></td>
<td>Calculate path lengths from the tree root</td></tr>
<tr><td style="width: 25%;"><a href="getPositionQuality.html">getPositionQuality</a></td>
<td>Get a data.frame with sequencing qualities per position</td></tr>
<tr><td style="width: 25%;"><a href="getSegment.html">getSegment</a></td>
<td>Get Ig segment allele, gene and family names</td></tr>
<tr><td style="width: 25%;"><a href="graphToPhylo.html">graphToPhylo</a></td>
<td>Convert a tree in igraph 'graph' format to ape 'phylo' format.</td></tr>
<tr><td style="width: 25%;"><a href="gravy.html">gravy</a></td>
<td>Calculates the hydrophobicity of amino acid sequences</td></tr>
<tr><td style="width: 25%;"><a href="gridPlot.html">gridPlot</a></td>
<td>Plot multiple ggplot objects</td></tr>
<tr><td style="width: 25%;"><a href="groupGenes.html">groupGenes</a></td>
<td>Group sequences by gene assignment</td></tr>
</table>

<h2><a name="I">-- I --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="DEFAULT_COLORS.html">IG_COLORS</a></td>
<td>Default colors</td></tr>
<tr><td style="width: 25%;"><a href="IMGT_REGIONS.html">IMGT_REGIONS</a></td>
<td>IMGT V-segment regions</td></tr>
<tr><td style="width: 25%;"><a href="isValidAASeq.html">isValidAASeq</a></td>
<td>Validate amino acid sequences</td></tr>
<tr><td style="width: 25%;"><a href="IUPAC_CODES.html">IUPAC_AA</a></td>
<td>IUPAC ambiguous characters</td></tr>
<tr><td style="width: 25%;"><a href="IUPAC_CODES.html">IUPAC_CODES</a></td>
<td>IUPAC ambiguous characters</td></tr>
<tr><td style="width: 25%;"><a href="IUPAC_CODES.html">IUPAC_DNA</a></td>
<td>IUPAC ambiguous characters</td></tr>
</table>

<h2><a name="J">-- J --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="junctionAlignment.html">junctionAlignment</a></td>
<td>Calculate junction region alignment properties</td></tr>
</table>

<h2><a name="M">-- M --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="makeChangeoClone.html">makeChangeoClone</a></td>
<td>Generate a ChangeoClone object for lineage construction</td></tr>
<tr><td style="width: 25%;"><a href="makeTempDir.html">makeTempDir</a></td>
<td>Create a temporary folder</td></tr>
<tr><td style="width: 25%;"><a href="maskPositionsByQuality.html">maskPositionsByQuality</a></td>
<td>Mask sequence positions with low quality</td></tr>
<tr><td style="width: 25%;"><a href="maskSeqEnds.html">maskSeqEnds</a></td>
<td>Masks ragged leading and trailing edges of aligned DNA sequences</td></tr>
<tr><td style="width: 25%;"><a href="maskSeqGaps.html">maskSeqGaps</a></td>
<td>Masks gap characters in DNA sequences</td></tr>
<tr><td style="width: 25%;"><a href="MRCATest-class.html">MRCATest</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="MRCATest-class.html">MRCATest-class</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="MRCATest-class.html">MRCATest-method</a></td>
<td>S4 class defining edge significance</td></tr>
</table>

<h2><a name="N">-- N --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="nonsquareDist.html">nonsquareDist</a></td>
<td>Calculate pairwise distances between sequences</td></tr>
</table>

<h2><a name="P">-- P --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="padSeqEnds.html">padSeqEnds</a></td>
<td>Pads ragged ends of aligned DNA sequences</td></tr>
<tr><td style="width: 25%;"><a href="pairwiseDist.html">pairwiseDist</a></td>
<td>Calculate pairwise distances between sequences</td></tr>
<tr><td style="width: 25%;"><a href="pairwiseEqual.html">pairwiseEqual</a></td>
<td>Calculate pairwise equivalence between sequences</td></tr>
<tr><td style="width: 25%;"><a href="permuteLabels.html">permuteLabels</a></td>
<td>Permute the node labels of a tree</td></tr>
<tr><td style="width: 25%;"><a href="phyloToGraph.html">phyloToGraph</a></td>
<td>Convert a tree in ape 'phylo' format to igraph 'graph' format.</td></tr>
<tr><td style="width: 25%;"><a href="AbundanceCurve-class.html">plot-method</a></td>
<td>S4 class defining a clonal abundance curve</td></tr>
<tr><td style="width: 25%;"><a href="DiversityCurve-class.html">plot-method</a></td>
<td>S4 class defining a diversity curve</td></tr>
<tr><td style="width: 25%;"><a href="EdgeTest-class.html">plot-method</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="MRCATest-class.html">plot-method</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="plotAbundanceCurve.html">plotAbundanceCurve</a></td>
<td>Plots a clonal abundance distribution</td></tr>
<tr><td style="width: 25%;"><a href="plotDiversityCurve.html">plotDiversityCurve</a></td>
<td>Plot the results of alphaDiversity</td></tr>
<tr><td style="width: 25%;"><a href="plotDiversityTest.html">plotDiversityTest</a></td>
<td>Plot the results of diversity testing</td></tr>
<tr><td style="width: 25%;"><a href="plotEdgeTest.html">plotEdgeTest</a></td>
<td>Plot the results of an edge permutation test</td></tr>
<tr><td style="width: 25%;"><a href="plotMRCATest.html">plotMRCATest</a></td>
<td>Plot the results of a founder permutation test</td></tr>
<tr><td style="width: 25%;"><a href="plotSubtrees.html">plotSubtrees</a></td>
<td>Plots subtree statistics for multiple trees</td></tr>
<tr><td style="width: 25%;"><a href="polar.html">polar</a></td>
<td>Calculates the average polarity of amino acid sequences</td></tr>
<tr><td style="width: 25%;"><a href="AbundanceCurve-class.html">print-method</a></td>
<td>S4 class defining a clonal abundance curve</td></tr>
<tr><td style="width: 25%;"><a href="DiversityCurve-class.html">print-method</a></td>
<td>S4 class defining a diversity curve</td></tr>
<tr><td style="width: 25%;"><a href="EdgeTest-class.html">print-method</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="MRCATest-class.html">print-method</a></td>
<td>S4 class defining edge significance</td></tr>
<tr><td style="width: 25%;"><a href="progressBar.html">progressBar</a></td>
<td>Standard progress bar</td></tr>
</table>

<h2><a name="R">-- R --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="rarefyDiversity.html">rarefyDiversity</a></td>
<td>Generate a clonal diversity index curve</td></tr>
<tr><td style="width: 25%;"><a href="readChangeoDb.html">readChangeoDb</a></td>
<td>Read a Change-O tab-delimited database file</td></tr>
<tr><td style="width: 25%;"><a href="readFastqDb.html">readFastqDb</a></td>
<td>Load sequencing quality scores from a FASTQ file</td></tr>
<tr><td style="width: 25%;"><a href="readIgphyml.html">readIgphyml</a></td>
<td>Read in output from IgPhyML</td></tr>
</table>

<h2><a name="S">-- S --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="seqDist.html">seqDist</a></td>
<td>Calculate distance between two sequences</td></tr>
<tr><td style="width: 25%;"><a href="seqEqual.html">seqEqual</a></td>
<td>Test DNA sequences for equality.</td></tr>
<tr><td style="width: 25%;"><a href="SingleDb.html">SingleDb</a></td>
<td>Single sequence AIRR database</td></tr>
<tr><td style="width: 25%;"><a href="sortGenes.html">sortGenes</a></td>
<td>Sort V(D)J genes</td></tr>
<tr><td style="width: 25%;"><a href="stoufferMeta.html">stoufferMeta</a></td>
<td>Weighted meta-analysis of p-values via Stouffer's method</td></tr>
<tr><td style="width: 25%;"><a href="summarizeSubtrees.html">summarizeSubtrees</a></td>
<td>Generate subtree summary statistics for a tree</td></tr>
</table>

<h2><a name="T">-- T --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="tableEdges.html">tableEdges</a></td>
<td>Tabulate the number of edges between annotations within a lineage tree</td></tr>
<tr><td style="width: 25%;"><a href="testDiversity.html">testDiversity</a></td>
<td>Pairwise test of the diversity index</td></tr>
<tr><td style="width: 25%;"><a href="testEdges.html">testEdges</a></td>
<td>Tests for parent-child annotation enchrichment in lineage trees</td></tr>
<tr><td style="width: 25%;"><a href="testMRCA.html">testMRCA</a></td>
<td>Tests for MRCA annotation enrichment in lineage trees</td></tr>
<tr><td style="width: 25%;"><a href="translateDNA.html">translateDNA</a></td>
<td>Translate nucleotide sequences to amino acids</td></tr>
<tr><td style="width: 25%;"><a href="translateStrings.html">translateStrings</a></td>
<td>Translate a vector of strings</td></tr>
<tr><td style="width: 25%;"><a href="DEFAULT_COLORS.html">TR_COLORS</a></td>
<td>Default colors</td></tr>
</table>

<h2><a name="W">-- W --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="writeChangeoDb.html">writeChangeoDb</a></td>
<td>Write a Change-O tab-delimited database file</td></tr>
</table>
</div></body></html>
