# SiCR: Web Application for Single Cell Repertoire Analysis
## 1. Introduction
SiCR is web application specialized for single cell repertoire analysis. 
## 2. Install
Download SiCR's docker image.

```bash
 # Download the docker image:
 docker pull 〇〇
```
## 3. How to use
### 3.1 One sample
#### 3.1.1 Cellranger
When you got fastq files generated by 10x Genomics Chromium Immune Profiling system, run cellranger multi. 
```bash
cellranger multi --id=YOUR_ID --csv=YOUR_CSV 
```
Please refer [cellranger website](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) more details.
#### 3.1.2 Run SiCR using cellranger outputs
For running SiCR, we need maximum three files generated by cellranger multi.
1. **filtered_feature_bc_matrix.h5** (mandatory)
<p>This is a count file written in hdf5 format. This file exists in YOUR_ID/outs/count directory

2. **filtered_contig_annotations.csv** for TCR (optional)
<p>When you have data for TCR, please find this file. This file exists in YOUR_ID/outs/vdj_t directory

3. **filtered_contig_annotations.csv** for BCR (optional)
<p>When you have data for BCR, please find this file. This file exists in YOUR_ID/outs/vdj_b directory




### 3.2 Multiple samples
#### 3.2.1 Cellranger
