<!DOCTYPE html>
<html>
<head><title>R: B Cell Receptor Phylogenetics Toolkit</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> B Cell Receptor Phylogenetics Toolkit
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;dowser&rsquo; version 1.1.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="airrClone-class.html">airrClone</a></td>
<td>S4 class defining a clone in Dowser</td></tr>
<tr><td style="width: 25%;"><a href="airrClone-class.html">airrClone-class</a></td>
<td>S4 class defining a clone in Dowser</td></tr>
<tr><td style="width: 25%;"><a href="BiopsyTrees.html">BiopsyTrees</a></td>
<td>Example Ig lineage trees with biopsy reconstructions.</td></tr>
<tr><td style="width: 25%;"><a href="bootstrapTrees.html">bootstrapTrees</a></td>
<td>Deprecated! Please use findSwitches instead.</td></tr>
<tr><td style="width: 25%;"><a href="buildClonalGermline.html">buildClonalGermline</a></td>
<td>'buildClonalGermline' Determine consensus clone sequence and create germline for clone</td></tr>
<tr><td style="width: 25%;"><a href="buildGermline.html">buildGermline</a></td>
<td>'buildGermline' reconstruct germline segments from alignment data</td></tr>
<tr><td style="width: 25%;"><a href="buildIgphyml.html">buildIgphyml</a></td>
<td>Wrapper to build IgPhyML trees and infer intermediate nodes</td></tr>
<tr><td style="width: 25%;"><a href="buildPhylo.html">buildPhylo</a></td>
<td>Wrapper for alakazam::buildPhylipLineage</td></tr>
<tr><td style="width: 25%;"><a href="buildPML.html">buildPML</a></td>
<td>Wrapper for phangorn::optim.pml</td></tr>
<tr><td style="width: 25%;"><a href="buildPratchet.html">buildPratchet</a></td>
<td>Wrapper for phangorn::pratchet</td></tr>
<tr><td style="width: 25%;"><a href="collapseNodes.html">collapseNodes</a></td>
<td>Collapse internal nodes with the same predicted sequence</td></tr>
<tr><td style="width: 25%;"><a href="colorTrees.html">colorTrees</a></td>
<td>Get a color palette for a predefined set of trait values</td></tr>
<tr><td style="width: 25%;"><a href="condenseTrees.html">condenseTrees</a></td>
<td>Condense a set of equally parsimonious node labels into a single tree</td></tr>
<tr><td style="width: 25%;"><a href="correlationTest.html">correlationTest</a></td>
<td>Run date randomization test for temporal signal on a set of trees.</td></tr>
<tr><td style="width: 25%;"><a href="createGermlines.html">createGermlines</a></td>
<td>createGermlines Determine consensus clone sequence and create germline for clone</td></tr>
<tr><td style="width: 25%;"><a href="downsampleClone.html">downsampleClone</a></td>
<td>'downsampleClone' Down-sample clone to maximum tip/switch ratio</td></tr>
<tr><td style="width: 25%;"><a href="dowser.html">dowser</a></td>
<td>The dowser package</td></tr>
<tr><td style="width: 25%;"><a href="ExampleAirr.html">ExampleAirr</a></td>
<td>Example AIRR database</td></tr>
<tr><td style="width: 25%;"><a href="ExampleClones.html">ExampleClones</a></td>
<td>Example Ig lineage trees</td></tr>
<tr><td style="width: 25%;"><a href="ExampleDbChangeo.html">ExampleDbChangeo</a></td>
<td>Example Change-O database</td></tr>
<tr><td style="width: 25%;"><a href="findSwitches.html">findSwitches</a></td>
<td>Create a bootstrap distribution for clone sequence alignments, and estimate trees for each bootstrap replicate.</td></tr>
<tr><td style="width: 25%;"><a href="formatClones.html">formatClones</a></td>
<td>Generate an ordered list of airrClone objects for lineage construction</td></tr>
<tr><td style="width: 25%;"><a href="getDivergence.html">getDivergence</a></td>
<td>Get divergence from root of tree for each tip</td></tr>
<tr><td style="width: 25%;"><a href="getGermline.html">getGermline</a></td>
<td>getGermline get germline segment from specified receptor and segment</td></tr>
<tr><td style="width: 25%;"><a href="getNodeSeq.html">getNodeSeq</a></td>
<td>Return IMGT gapped sequence of specified tree node</td></tr>
<tr><td style="width: 25%;"><a href="getPalette.html">getPalette</a></td>
<td>Get a color palette for a predefined set of trait values</td></tr>
<tr><td style="width: 25%;"><a href="getSeq.html">getSeq</a></td>
<td>Deprecated! Use getNodeSeq</td></tr>
<tr><td style="width: 25%;"><a href="getSubclones.html">getSubclones</a></td>
<td>Define subclones based on light chain rearrangements</td></tr>
<tr><td style="width: 25%;"><a href="getTrees.html">getTrees</a></td>
<td>Estimate lineage tree topologies, branch lengths, and internal node states if desired</td></tr>
<tr><td style="width: 25%;"><a href="IsotypeTrees.html">IsotypeTrees</a></td>
<td>Example Ig lineage trees with isotype reconstructions.</td></tr>
<tr><td style="width: 25%;"><a href="makeAirrClone.html">makeAirrClone</a></td>
<td>Generate a airrClone object for lineage construction</td></tr>
<tr><td style="width: 25%;"><a href="makeModelFile.html">makeModelFile</a></td>
<td>Make a parsimony model file</td></tr>
<tr><td style="width: 25%;"><a href="maskCodons.html">maskCodons</a></td>
<td>'maskCodons' Masks codons split by insertions</td></tr>
<tr><td style="width: 25%;"><a href="maskSequences.html">maskSequences</a></td>
<td>'maskSequences' Mask codons split by insertions in V gene</td></tr>
<tr><td style="width: 25%;"><a href="plotTrees.html">plotTrees</a></td>
<td>Plot a tree with colored internal node labels using ggtree</td></tr>
<tr><td style="width: 25%;"><a href="readFasta.html">readFasta</a></td>
<td>Read a fasta file into a list of sequences 'readFasta' reads a fasta file</td></tr>
<tr><td style="width: 25%;"><a href="readIMGT.html">readIMGT</a></td>
<td>'readIMGT' read in IMGT database</td></tr>
<tr><td style="width: 25%;"><a href="readLineages.html">readLineages</a></td>
<td>Read in all trees from a lineages file</td></tr>
<tr><td style="width: 25%;"><a href="readModelFile.html">readModelFile</a></td>
<td>Read in a parsimony model file</td></tr>
<tr><td style="width: 25%;"><a href="reconIgPhyML.html">reconIgPhyML</a></td>
<td>Do IgPhyML maximum parsimony reconstruction</td></tr>
<tr><td style="width: 25%;"><a href="rerootTree.html">rerootTree</a></td>
<td>Reroot phylogenetic tree to have its germline sequence at a zero-length branch to a node which is the direct ancestor of the tree's UCA. Assigns 'uca' to be the ancestral node to the tree's germline sequence, as 'germid' as the tree's germline sequence ID.</td></tr>
<tr><td style="width: 25%;"><a href="resolvePolytomies.html">resolvePolytomies</a></td>
<td>Resolve polytomies to have the minimum number of single timepoint clades</td></tr>
<tr><td style="width: 25%;"><a href="runCorrelationTest.html">runCorrelationTest</a></td>
<td>Run correlationTest, based on https://doi.org/10.1111/2041-210X.12466</td></tr>
<tr><td style="width: 25%;"><a href="scaleBranches.html">scaleBranches</a></td>
<td>Scale branch lengths to represent either mutations or mutations per site.</td></tr>
<tr><td style="width: 25%;"><a href="stitchRegions.html">stitchRegions</a></td>
<td>stitchRegions Similar to stitchVDJ but with segment IDs instead of nulecotides</td></tr>
<tr><td style="width: 25%;"><a href="stitchVDJ.html">stitchVDJ</a></td>
<td>stitchVDJ combines germline gene segments to a single string</td></tr>
<tr><td style="width: 25%;"><a href="testPS.html">testPS</a></td>
<td>Performs PS (parsimony score) test on switch data</td></tr>
<tr><td style="width: 25%;"><a href="testSC.html">testSC</a></td>
<td>Performs SC (switch count) test on switch data</td></tr>
<tr><td style="width: 25%;"><a href="testSP.html">testSP</a></td>
<td>Performs SP (switch proportion) test on switch data</td></tr>
<tr><td style="width: 25%;"><a href="TimeTrees.html">TimeTrees</a></td>
<td>Example Ig lineage trees sampled over time.</td></tr>
<tr><td style="width: 25%;"><a href="treesToPDF.html">treesToPDF</a></td>
<td>Simple function for plotting a lot of trees into a pdf</td></tr>
<tr><td style="width: 25%;"><a href="writeLineageFile.html">writeLineageFile</a></td>
<td>Write lineage file for IgPhyML use</td></tr>
</table>
</div></body></html>
